src/ssh/test-sshbridge.c

```diff
@@ -0,0 +1,1358 @@
+        {
+          return input;
+        }
+      else
+        {
+          input->len = len + ret;
+          input->str[input->len] = '\0';
+        }
+    }
+}
+
+static void
+spawn_setup (gpointer data)
+{
+  int fd = GPOINTER_TO_INT (data);
+
+  /* Send this signal to all direct child processes, when bridge dies */
+  prctl (PR_SET_PDEATHSIG, SIGHUP);
+
+  if (dup2 (fd, 0) < 0 || dup2 (fd, 1) < 0)
+    g_assert_not_reached ();
+
+  close (fd);
+}
+
+static void
+setup_mock_sshd (TestCase *tc,
+                 gconstpointer data)
+{
+  const TestFixture *fixture = data;
+  GError *error = NULL;
+  GString *port;
+  gchar *endptr;
+  guint64 value;
+  gint out_fd;
+
+  const gchar *argv[] = {
+      BUILDDIR "/mock-sshd",
+      "--user", g_get_user_name (),
+      "--password", PASSWORD,
+      fixture->mock_sshd_arg,
+      NULL
+  };
+
+  g_spawn_async_with_pipes (BUILDDIR, (gchar **)argv, NULL, G_SPAWN_DO_NOT_REAP_CHILD, NULL, NULL,
+                            &tc->mock_sshd, NULL, &out_fd, NULL, &error);
+  g_assert_no_error (error);
+
+  /*
+   * mock-sshd prints its port on stdout, and then closes stdout
+   * This also lets us know when it has initialized.
+   */
+
+  port = read_all_into_string (out_fd);
+  g_assert (port != NULL);
+  close (out_fd);
+  g_assert_no_error (error);
+
+  g_strstrip (port->str);
+  value = g_ascii_strtoull (port->str, &endptr, 10);
+  if (!endptr || *endptr != '\0' || value == 0 || value > G_MAXUSHORT)
+      g_critical ("invalid port printed by mock-sshd: %s", port->str);
+
+  tc->ssh_port = (gushort)value;
+  g_string_free (port, TRUE);
+}
+
+static const TestFixture fixture_mock_echo = {
+  .ssh_command = BUILDDIR "/mock-echo"
+};
+
+static const TestFixture fixture_cat = {
+  .ssh_command =  SRCDIR "/src/ws/mock-cat-with-init"
+};
+
+static gchar **
+setup_env (const TestFixture *fix)
+{
+  const gchar *command;
+  const gchar *knownhosts_file;
+  const gchar *config;
+  gchar **env = g_get_environ ();
+
+  config = fix ? fix->config : NULL;
+  if (!config)
+    config = SRCDIR "/src/ssh/mock-config";
+  env = g_environ_setenv (env, "XDG_CONFIG_DIRS", config, TRUE);
+
+  command = fix ? fix->ssh_command : NULL;
+  if (!command)
+    command = fixture_cat.ssh_command;
+  env = g_environ_setenv (env, "COCKPIT_SSH_BRIDGE_COMMAND", command, TRUE);
+
+  if (fix && fix->allow_unknown)
+    {
+      env = g_environ_setenv (env, "COCKPIT_SSH_ALLOW_UNKNOWN",
+                              "true", TRUE);
+    }
+
+  knownhosts_file = fix ? fix->knownhosts_file : NULL;
+  if (!knownhosts_file)
+      knownhosts_file = SRCDIR "/src/ssh/mock_known_hosts";
+
+  env = g_environ_setenv (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE",
+                          knownhosts_file, TRUE);
+  return env;
+}
+
+static CockpitTransport *
+start_bridge (gchar **env,
+              gchar **argv)
+{
+  GError *error = NULL;
+  int fds[2];
+
+  if (socketpair (PF_LOCAL, SOCK_STREAM, 0, fds) < 0)
+    g_assert_not_reached ();
```

Comment:
Nitpick: I think `g_assert_cmpint(..., >=, 0)` is both more concise and clear, and provides more useful failure messages as it would show the errno. This Should Not Happen™ though. (This pattern is also being used several times here)